lib/src/gpg_signing.rs

```diff
@@ -0,0 +1,232 @@
+
+#![allow(missing_docs)]
+
+use std::ffi::{OsStr, OsString};
+use std::fmt::Debug;
+use std::io::Write;
+use std::process::{Command, ExitStatus, Stdio};
+use std::str;
+
+use thiserror::Error;
+
+use crate::signing::{SigStatus, SignError, SigningBackend, Verification};
+
+// Search for one of the:
+//  [GNUPG:] GOODSIG <long keyid> <primary uid..>
+//  [GNUPG:] EXPKEYSIG <long keyid> <primary uid..>
+//  [GNUPG:] NO_PUBKEY <long keyid>
+//  [GNUPG:] BADSIG <long keyid> <primary uid..>
+// in the output from --status-fd=1
+// Assume signature is invalid if none of the above was found
+fn parse_gpg_verify_output(
+    output: &[u8],
+    allow_expired_keys: bool,
+) -> Result<Verification, SignError> {
+    output
+        .split(|&b| b == b'\n')
+        .filter_map(|line| line.strip_prefix(b"[GNUPG:] "))
+        .find_map(|line| {
+            let mut parts = line.splitn(3, |&b| b == b' ').fuse();
+            let status = match parts.next()? {
+                b"GOODSIG" => SigStatus::Good,
+                b"EXPKEYSIG" => {
+                    if allow_expired_keys {
+                        SigStatus::Good
+                    } else {
+                        SigStatus::Bad
+                    }
+                }
+                b"NO_PUBKEY" => SigStatus::Unknown,
+                b"BADSIG" => SigStatus::Bad,
+                _ => return None,
+            };
+            let key = parts
+                .next()
+                .and_then(|bs| str::from_utf8(bs).ok())
+                .map(|value| value.trim().to_owned());
+            let display = parts
+                .next()
+                .and_then(|bs| str::from_utf8(bs).ok())
+                .map(|value| value.trim().to_owned());
+            Some(Verification::new(status, key, display))
+        })
+        .ok_or(SignError::InvalidSignatureFormat)
+}
+
+#[derive(Debug)]
+pub struct GpgBackend {
+    program: OsString,
+    allow_expired_keys: bool,
+    extra_args: Vec<OsString>,
+}
+
+#[derive(Debug, Error)]
+pub enum GpgError {
+    #[error("GPG failed with exit status {exit_status}:\n{stderr}")]
+    Command {
+        exit_status: ExitStatus,
+        stderr: String,
+    },
+    #[error("Failed to run GPG")]
+    Io(#[from] std::io::Error),
+}
+
+impl From<GpgError> for SignError {
+    fn from(e: GpgError) -> Self {
+        SignError::Backend(Box::new(e))
+    }
+}
+
+impl GpgBackend {
+    pub fn new(program: OsString, allow_expired_keys: bool) -> Self {
+        Self {
+            program,
+            allow_expired_keys,
+            extra_args: vec![],
+        }
+    }
+
+    /// Primarily intended for testing
+    pub fn with_extra_args(mut self, args: &[OsString]) -> Self {
+        self.extra_args.extend_from_slice(args);
+        self
+    }
+
+    pub fn from_config(config: &config::Config) -> Self {
+        Self::new(
+            config
+                .get_string("signing.backends.gpg.program")
+                .unwrap_or_else(|_| "gpg2".into())
```

Comment:
@necauqua maybe the default should be "gpg"?
(just spotted because the test uses "gpg" not "gpg2")

Reply:
I think the gpg2 command is the one that requires the agent to work?.
So that agent use is enforced since we sign every snapshot and whatnot

Anyway I don't remember there being any other reasons, so could be gpg, yes, maybe I forgot to get back to think about this

Reply:
Perhaps, it's historical thing? I don't have `gpg2` in my Debian unstable environment, and `gpg --version` says it's 2.x.
https://packages.debian.org/bookworm/gnupg2

Reply:
You don't?.
I thought it was a symlink that every gpg 2.x also installs or something
I have it on NixOS
Yes it is a historical thing afaiu

Anyway I've no objections to changing the default to just gpg
Any unusual or historical setups could change the config value anyway